### ISSUE TITLE
Segmentation fault, how to troubleshoot? 2k small task problem [tune]

### ISSUE THREAD

**Issue by mike576** [ray-project/ray#13881]
using ray: version 2.0.0dev0 (same problem with 1.1.0 )
python: 3.7
os: linux ubuntu 18. (same on local mac os catalina)
Using anaconda for both. 

### What is the problem?

We build a lot of models 20k-100k, sometimes only 1k with grid search, so we have 20k-100k trials too. We don't want to find the best model, we‘d like to look at the distribution of the ML models and their performance (eg. distribution of r2 ). To train a model, it takes only about 0.2-0.5 sec, so you can consider it as a small task but not as a tiny task. So a week ago we started to use Tune and Ray with Jupyter notebook, on a bigger linux machine on aws with 96 CPUs, no other nodes, and like 75% of the runs the **kernel crashes OR the run hangs at the very last trial and does nothing**. I cannot find any log which would describe the cause of the crash.
I do understand that there is some overhead with starting trials and also copy the objects to the object store etc. I have read in another thread the tiny task problem too.

As I have found out the 100 trials run perfectly, so there's must be something wrong with 20k-100k. 

In htop, I can see a lot of IDLE ray processes but the machine is utilized about 100%, so it looks good. Note: we set checkpoint_freq=500 (avoiding saving checkpoints every 10th trial.)

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data/environments):

It MUST have external library dependencies but only> pandas numpy scikit-learn more-itertools

So running this before the code helps> 
`pip install pandas numpy scikit-learn more-itertools ray[tune]`

CODE:

```

import pandas as pd
import numpy as np
import ray
from ray import tune
from itertools import combinations


#Note I run it on aws c5.24xlarge with 96 CPU
if ray.is_initialized() == False:
    import multiprocessing
    num_cores = multiprocessing.cpu_count()
    ray.init(num_cpus=num_cores)



def run_for_one_param(params):
    from sklearn.linear_model import LinearRegression
    from sklearn.metrics import r2_score

    # I know this is a but ugly, agdata_ref should not be in params, but we have multiple of these dynamic paramters.
    agdata=ray.get(params["agdata_ref"])
    df=agdata.data
    y=df.pop("y")
    X=df
    linreg=LinearRegression()
    #I should use train test validation split I know...
    linreg.fit(X,y)
    pred_y=linreg.predict(X)

    report_dict = dict()
    report_dict["score_r2"]=r2_score(y,pred_y)
    return report_dict

def run_for_one_param_and_yield(params):
    yield run_for_one_param(params)

class AGData():

    def __init__(self):
        self.storage_dict=dict()

    def set_to_storage(self,key,value):
        self.storage_dict[key] = value

    def get_from_storage(self,key):
        return self.storage_dict[key]

    def set_data(self,data):
        self.data=data

    pass

def generate_combinations(features,min_len=1,max_len=10):
    all_comb=[]
    from itertools import combinations
    for i in range(min_len-1,max_len):
        comb=combinations(features, i+1)
        for act_comb in list(comb):
            all_comb.append(list(act_comb))

    return all_comb

class Runray():
    def __init__(self):
        self.agdata=AGData()
        #Add here new letters to grow trial numbers!!!!!
        self.colums='ABCDEFGHIJKL'
        df = pd.DataFrame(np.random.randint(0,100,size=(100, len(self.colums)+1)), columns=list(self.colums+"y"))
        self.agdata.set_data(df)

    def get_search_space_ray(self):

        combinations=generate_combinations(list(self.colums))
        print("nr trial will run: ",len(combinations))
        search_space = {
            'col_combinations': tune.grid_search(combinations),
            #'other params': tune.grid_search(some_list)
        }

        agdata_ref=ray.put(self.agdata)
        print("new agdata reference put into params.",agdata_ref)
        # I know this is totaly ugly, but this is how we pass the reference now:
        search_space["agdata_ref"]=tune.choice([agdata_ref])
        return search_space

    def run_ray(self):
        # RUN RAY
        analysis = tune.run(
            run_for_one_param_and_yield
            , config=self.get_search_space_ray()
            , num_samples=1
            ,verbose=1
            ,checkpoint_freq=500
            ,checkpoint_at_end=False
            ,keep_checkpoints_num=10
        )

        print("ray.version: ",ray.__version__)
        print(analysis.results_df)


Runray().run_ray()


```

This is the end of console Output:
```
...
== Status ==
Memory usage on this node: 5.3/184.7 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/96 CPUs, 0/0 GPUs, 0.0/119.19 GiB heap, 0.0/37.99 GiB objects
Result logdir: /home/ubuntu/ray_results/run_for_one_param_and_yield_2021-02-03_10-19-49
Number of trials: 2046/2046 (2046 TERMINATED)

2021-02-03 10:20:49,335 INFO tune.py:461 -- Total run time: 59.91 seconds (58.70 seconds for the tuning loop).
**Segmentation fault (core dumped)**
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue:
Here I tried and it runs perfectly with a clean environment. 
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).
Yes, segmentation fault also in ray 2.0.0dev0.
Segmentation fault also in ray 1.1.0

So basically there must be something wrong with underlying libraries. Basicly we use nbdev, and some other libs. Here are the requirements> 
With this env we got the segmentation fault:
[requirements_our_dev_system.txt](https://github.com/ray-project/ray/files/5917977/requirements_our_dev_system.txt)
With this env we got NO problem:
[requirements_clean_env_few_libs.txt](https://github.com/ray-project/ray/files/5917978/requirements_clean_env_few_libs.txt)

![image](https://user-images.githubusercontent.com/1930858/106739204-266faa80-6619-11eb-9fca-8d978cd5a6f7.png)

THANKS for ANY HELP!
Miklos

**Comment by rkooo567** [comment]
Can you check /tmp/ray/session_latest/logs and see which log file has the segmentation fault?

**Comment by rkooo567** [comment]
Also what’s ur cluster size? (How many nodes & cpus)?

**Comment by rkooo567** [comment]
Also can you monitor the memory usage while you are running ur stuff with 20+k trials?

**Comment by rkooo567** [comment]
Also cc @amogkam

**Comment by mike576** [comment]
@rkooo567 , great questions. Thanks for helping out. 

> Can you check /tmp/ray/session_latest/logs and see which log file has the segmentation fault?

![image](https://user-images.githubusercontent.com/1930858/106810335-1681b600-666d-11eb-8bf1-9e1f92fc9181.png)
![image](https://user-images.githubusercontent.com/1930858/106811758-1682b580-666f-11eb-8215-1170d50235c9.png)
![image](https://user-images.githubusercontent.com/1930858/106811790-239fa480-666f-11eb-8673-59fea357c1f5.png)
Dashboard has some errors>
![image](https://user-images.githubusercontent.com/1930858/106812098-86913b80-666f-11eb-836b-b2eb64eab013.png)


There is nothing like "Segmentation" or "fault" that in the log dir. BTW, there are 15k files in that log dir, isnt it a problem? I mean for an underlying process to open and handle a lot of files. Just guessing.

> Also what’s ur cluster size? (How many nodes & cpus)?

1 node only with 96 CPU. Aws c5.24xlarge. 196 GB Mem. 

> Also can you monitor the memory usage while you are running ur stuff with 20+k trials?

Monitored nothing strange there:
![image](https://user-images.githubusercontent.com/1930858/106810578-7710f300-666d-11eb-901c-c349a14edb2e.png)
![image](https://user-images.githubusercontent.com/1930858/106810608-842de200-666d-11eb-86f7-9416c9c69817.png)

**Comment by mike576** [comment]
Just wanted to mention that not every run ends up with segmentation fault. Especially with a fresh new start of the OS. I just did a test with 4k trials. The first 3 runs went well then 1 failed, then 1 went OK, then failed, failed, failed.

**Comment by rkooo567** [comment]
Hmm that's interesting. @amogkam Is there any issue with the configuration? also can you guys take a look at this issue next week?

Also, @mike576 canyou post a log zip file here?

**Comment by amogkam** [comment]
Hey @mike576 thanks for raising this issue. This is odd behvaior- I've never seen this segfault before. Also nothing stands in the code as incorrect. @krfricke did you run into  anything similar during your scalability experiments with large number of trials?

Looking at your requirements files, there are differences in `cloudpickle`, `numpy` and `pyarrow`. These could all potentially have an effect here. Do you think you can change just these libraries to the versions that work?

Also, can you try a couple things to see if they help:
- Removing the 3 checkpoint related arguments.
- Try outside of a Jupyter notebook (just ssh into the node and then run the python script).

Let us know if any of these help, and hopefully that can help identify the problem. We'll also try to reproduce this on our end next week!

**Comment by krfricke** [comment]
I've never seen this error before, either, but I only scaled to 10k trials.

So the interesting part here is that it prints the runtime but it does not print the commands that come after your call to `tune.run()`. The only thing that happens in between these two outputs is that the experiment analysis object is created and returned. 

Could you try to run this in a separate script?

```
from ray import tune

out = tune.ExperimentAnalysis(
    "~/ray_results/run_for_one_param_and_yield_2021-02-03_10-05-01")
print(out)
```
(maybe change the experiment dir accordingly)

Just to see if it crashed again. I don't suspect it will, as I think it's related to some form of resource exhaustion, but it would be great to see this just in case.

Also, can you give us the output of `du -hs ~/ray_results/run_for_one_param_and_yield_2021-02-03_10-05-01/*`?

**Comment by mike576** [comment]
> I've never seen this error before, either, but I only scaled to 10k trials.
> 
> So the interesting part here is that it prints the runtime but it does not print the commands that come after your call to `tune.run()`. The only thing that happens in between these two outputs is that the experiment analysis object is created and returned.
> 
> Could you try to run this in a separate script?
> 
> ```
> from ray import tune
> 
> out = tune.ExperimentAnalysis(
>     "~/ray_results/run_for_one_param_and_yield_2021-02-03_10-05-01")
> print(out)
> ```
> 
> (maybe change the experiment dir accordingly)


So, that tune.ExperimentAnalysis kills the kernel again
![image](https://user-images.githubusercontent.com/1930858/106958072-26ff6280-6739-11eb-89a5-f426e9ead487.png)

> 
> Just to see if it crashed again. I don't suspect it will, as I think it's related to some form of resource exhaustion, but it would be great to see this just in case.
> 
> Also, can you give us the output of `du -hs ~/ray_results/run_for_one_param_and_yield_2021-02-03_10-05-01/*`?

[du_result.txt](https://github.com/ray-project/ray/files/5928692/du_result.txt)

**Comment by mike576** [comment]
[ray_regult_issue13881.zip](https://github.com/ray-project/ray/files/5928703/ray_regult_issue13881.zip)
You may need this result dir, @krfricke .

**Comment by mike576** [comment]
> Hey @mike576 thanks for raising this issue. This is odd behvaior- I've never seen this segfault before. Also nothing stands in the code as incorrect. @krfricke did you run into anything similar during your scalability experiments with large number of trials?
> 
> Looking at your requirements files, there are differences in `cloudpickle`, `numpy` and `pyarrow`. These could all potentially have an effect here. Do you think you can change just these libraries to the versions that work?

Thanks. good points! I will test tomorrow these libs.

> 
> Also, can you try a couple things to see if they help:
> 
> * Removing the 3 checkpoint related arguments.

Sadly, same error by removing chkpt args.

> * Try outside of a Jupyter notebook (just ssh into the node and then run the python script).
>

Tried, same error...
 
> Let us know if any of these help, and hopefully that can help identify the problem. We'll also try to reproduce this on our end next week!

**Comment by mike576** [comment]
> Hmm that's interesting. @amogkam Is there any issue with the configuration? also can you guys take a look at this issue next week?
> 
> Also, @mike576 can you post a log zip file here?

THanks, here is the log.
[ray_result_log_issue13881.zip](https://github.com/ray-project/ray/files/5928791/ray_result_log_issue13881.zip)

**Comment by mike576** [comment]
Hi @amogkam @krfricke @rkooo567 ,
Absolutely grateful for your help! I answered all your questions, uploaded everything I could. 
I think this might be a library version problem. I mean, the cause might be the wrong version of pyarrow, or cloudpickle. (I use anaconda)
Tomorrow, I will try to update the libs.

**Comment by krfricke** [comment]
So I can't reproduce the error - i.e. for me it loads the experiment analysis object correctly (and this is on a laptop). So it might be a library error after all. If you find anything there, can you share any insights?

**Comment by krfricke** [comment]
I'll close this for now - please feel free to re-open if the issue still comes up. Thanks!

**Comment by mike576** [comment]
Thanks, actually, I could not figure out what was wrong, with a clean environment it works.
